### PR TITLE
modified retrieve token by product api to return also the legals with…

### PR DESCRIPTION
**Pull request by KevinSi96**
… the token informations

#### List of Changes

The getTokensByProduct wasn't returning the legals associated to the onboarding token

#### Motivation and Context

External interceptor needs to retrieve the tokens given a product and send them to sap, with the token it has to send also the legals associated to it

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: connector-api/src/main/java/it/pagopa/selfcare/mscore/api/UserConnector.java [hunk 1]
```diff
@@ -50,13 +50,12 @@ public interface UserConnector {
 
     void findAndRemoveProduct(String userId, String institutionId, OnboardedProduct product);
 
-    List<OnboardedUser> findAllByIds(List<String> users);
+    List<OnboardedUser> findAllByIds(List<String> users, boolean existingOnly);
```
Comment (gunzip) [inline connector-api/src/main/java/it/pagopa/selfcare/mscore/api/UserConnector.java:53]: What about introducing a new method `findExistingProductsByIds()`? 1. this usually results a better dx than polluting arguments with booleans 2. you avoid to refactor all exiting instaces of `findAllByIds`. You can keep all the existing code as a base for the two different methods making it private.

File: core/src/main/java/it/pagopa/selfcare/mscore/core/TokenService.java [hunk 2]
```diff
@@ -11,10 +11,10 @@ public interface TokenService {
 
     Token verifyOnboarding(String institutionId, String productId);
 
-    TokenRelationships retrieveToken(String tokenId);
+    TokenRelationships retrieveToken(String tokenId, boolean existingOnly);
```
Comment (gunzip) [inline core/src/main/java/it/pagopa/selfcare/mscore/core/TokenService.java:14]: ditto

File: connector-api/src/main/java/it/pagopa/selfcare/mscore/api/UserConnector.java (hunk 1 above)
Comment (gunzip) [reply connector-api/src/main/java/it/pagopa/selfcare/mscore/api/UserConnector.java:53]: Generally it's better to not expose public interfaces with boolean arguments. When really needed use an option object instead.

File: connector/dao/src/main/java/it/pagopa/selfcare/mscore/connector/dao/UserConnectorImpl.java [hunk 3]
```diff
@@ -81,15 +81,15 @@ public OnboardedUser save(OnboardedUser onboardedUser) {
...
         repository.findAllById(users)
                 .forEach(userEntity -> {
                     userList.add(userMapper.toOnboardedUser(userEntity));
                     userIds.remove(userEntity.getId());
                 });
-        if (users.size() != userList.size()) {
+        if (users.size() != userList.size() && !existingOnly) {
             throw new ResourceNotFoundException(String.format(USERS_NOT_FOUND_ERROR.getMessage(), String.join(",", userIds)), USERS_NOT_FOUND_ERROR.getCode());
         }
         return userList;
```
Comment (manuraf) [inline connector/dao/src/main/java/it/pagopa/selfcare/mscore/connector/dao/UserConnectorImpl.java:95]: you could make it clearer, for example:

```
public List<OnboardedUser> findAllByIds(List<String> userIds, boolean existingOnly) {
    List<OnboardedUser> userList = new ArrayList<>();
    Set<String> remainingUserIds = new HashSet<>(userIds);
    
    repository.findAllById(userIds).forEach(userEntity -> {
        userList.add(userMapper.toOnboardedUser(userEntity));
        remainingUserIds.remove(userEntity.getId());
    });

    if (!existingOnly && !remainingUserIds.isEmpty()) {
        String missingUserIds = String.join(",", remainingUserIds);
        throw new ResourceNotFoundException(String.format(USERS_NOT_FOUND_ERROR.getMessage(), missingUserIds), USERS_NOT_FOUND_ERROR.getCode());
    }

    return userList;
}
```

File: core/src/main/java/it/pagopa/selfcare/mscore/core/TokenServiceImpl.java [hunk 4]
```diff
@@ -73,12 +73,14 @@ public Token getToken(String institutionId, String productId){
...
+        log.debug("retrieveToken tokenId = {}, existingOnly = {}", tokenId, existingOnly);
         Token token = tokenConnector.findById(tokenId);
         List<OnboardedUser> users;
         if (token.getUsers() != null) {
             var ids = token.getUsers().stream().map(TokenUser::getUserId).collect(Collectors.toList());
-            users = userService.findAllByIds(ids);
+                users = userService.findAllByIds(ids, existingOnly);
         } else {
             users = Collections.emptyList();
         }
```
Comment (manuraf) [inline core/src/main/java/it/pagopa/selfcare/mscore/core/TokenServiceImpl.java:86]: you should avoid the if, for example:

```
List<OnboardedUser> users = Optional.ofNullable(token.getUsers())
                  .map(users -> users.stream().map(TokenUser::getUserId).collect(Collectors.toList()))
                  .map(ids -> userService.findAllByIds(ids, existingOnly)
                  .orElse(Collections.emptyList())
```

File: connector-api/src/main/java/it/pagopa/selfcare/mscore/api/UserConnector.java (hunk 1 above)
Comment (manuraf) [reply connector-api/src/main/java/it/pagopa/selfcare/mscore/api/UserConnector.java:53]: you could introduce 2 method public (findAllByIds and indExistingProductsByIds) and using only one private method that implement the logic with boolean...

**Review by manuraf — CHANGES_REQUESTED** [review]

**Review by manuraf — APPROVED** [review]

**Pull request merged**